### PR TITLE
Extract methods

### PR DESCRIPTION
**Pull request by ziofil**
------------------------------------------------------------------------------------------------------------

**Description of the Change:**
Added two new utilities to extract a numerical representation of a circuit from an Engine object: `extract_unitary` and `extract_channel`.
The first returns a numerical array that represents a unitary transformation (and it's applicable to the case where in the circuit specification there are only ops of type `Gate`).
The second returns a numerical array that represents a channel (and it's applicable to the case where in the circuit specification there are only ops of type `Gate` or `Channel`).

Both functions support an option to vectorise the indices that correspond to different modes. This makes `extract_unitary` return a matrix, and `extract_channel` return a rank-4 tensor (if the option `representation` is 'choi' or 'liouville') or a rank-3 tensor (if the option `representation` is 'kraus')

`extract_unitary` also supports the option `backend` which can take the value 'fock' (default), which results in a numpy array output, or 'tf' which results in a TensorFlow tensor output.

**Benefits:**
Now a user can specify a circuit and ask for a numerical representation of the transformation that it represents.

**Possible Drawbacks:**
What drawbacks? xD

**Related GitHub Issues:**
None (that I know of)

### CODE REVIEW

File: strawberryfields/utils.py [hunk 1]
```diff
@@ -673,3 +675,239 @@ def f_proxy(*args):
...
+def is_unitary(engine):
+    "Returns true if every command in the queue is of type Gate"
+    return all(isinstance(cmd.op, Gate) for cmd in engine.cmd_queue)
+
+def is_channel(engine):
+    "Returns true if every command in the queue is either of type Gate or type Channel"
+    return all(isinstance(cmd.op, (Channel, Gate)) for cmd in engine.cmd_queue)
+
+def vectorize_dm(tensor):
+    """Given a tensor with 4N indices each of dimension D each,
```
Comment (ziofil) [inline strawberryfields/utils.py]: there's an "each" too many

**Comment by co9olguy** [comment]
Great, thanks for the PR @ziofil!

I notice some of the tests are currently failing (https://travis-ci.org/XanaduAI/strawberryfields/builds/498260170?utm_source=github_status&utm_medium=notification). Is this still a work-in-progress, or do you expect to resolve those shortly? If so, we can get started on code review.

**Comment by ziofil** [comment]
Hi Nathan,
Ah, it's because I used f-strings, and you're also checking that the code works for python 3.5
I'll update the code to use the `format()` syntax.

**Comment by josh146** [comment]
Ignore the codecov bot for now - the comment will be automatically updated with the correct coverage once the build passes

**Comment by ziofil** [comment]
The remaining issue is that I need to use the Engine's method `_add_subsystems()`, and it begins with an underscore... I don't see a way out other than refactor the code and make it part of the public interface, but it's probably even worse.

**Comment by josh146** [comment]
@ziofil, all good - we're okay with ignoring that linting rule :slightly_smiling_face:

Everything looks like it's passing now :champagne:

Will you be adding any more commits, or is it okay to begin the code review?

**Comment by ziofil** [comment]
No, I'm done. 👍

File: strawberryfields/utils.py [hunk 2]
```diff
@@ -673,3 +691,334 @@ def f_proxy(*args):
...
+
+    Here, CJ is from Choi-Jamiolkowski, as under the hood we are expliting the CJ isomorphism.
+
+    Args:
+        engine (Engine): a Strawberry Fields engine instance
+        cutoff_dim (int): the Fock basis truncation
+
+    Returns:
+        strawberryfields.engine.Engine: returns the same Engine instance, with the number of subsystems
+        increased and the interleaved identities ket prepended to the queue
```
Comment (josh146) [inline strawberryfields/utils.py]: Just double checking, but because there is no deepcopy here, this is an 'in-place' engine operation?

Comment (ziofil) [reply strawberryfields/utils.py]: Hi Josh, I'm not sure I understand your question. Could you rephrase?

File: strawberryfields/utils.py [hunk 3]
```diff
@@ -673,3 +691,334 @@ def f_proxy(*args):
...
+    elif representation.lower() == 'liouville':
+        result = np.einsum('abcd -> dbca', choi)
+        if not vectorize_modes:
+            result = _unvectorize_dm(result, N)
+
+    elif representation.lower() == 'kraus':
+        eigvals, eigvecs = np.linalg.eig(np.einsum('abcd -> cadb', choi).reshape([cutoff_dim**(2*N), cutoff_dim**(2*N)]))
+        eigvecs = eigvecs[:, ~np.isclose(abs(eigvals), 0)]
+        eigvals = eigvals[~np.isclose(abs(eigvals), 0)]
+        result = np.einsum('abc->cab', np.einsum('b,ab->ab', np.sqrt(eigvals), eigvecs).reshape([cutoff_dim**N, cutoff_dim**N, -1]))
```
Comment (josh146) [inline strawberryfields/utils.py]: @ziofil, your docstring is fantastic, but could you also add some comments here (especially to the Kraus lines) just to clarify what is happening on each line?

File: tests/test_utils.py [hunk 4]
```diff
@@ -290,13 +292,211 @@ def test_random_interferometer(self):
...
+        with tf.Session() as sess:
+            sess.run(tf.global_variables_initializer())
+            in_state = tf.constant(initial_state.reshape([-1]), dtype=tf.complex64)
+            final_state_tf_backend = sess.run(tf.einsum('ab,b', matrix_extract_tf_backend, in_state))
+        final_state_sf = eng_sf.run("fock", cutoff_dim=cutoff_dim).ket().reshape([-1])
+        final_state_extract = matrix_extract.dot(initial_state)
+        final_state_sf = eng_sf.run("fock", cutoff_dim=cutoff_dim).ket()
+
+        self.assertAllAlmostEqual(final_state_extract, final_state_sf, delta=defaults.TOLERANCE)
+        self.assertAllAlmostEqual(final_state_tf_backend, final_state_sf, delta=defaults.TOLERANCE)
```
Comment (josh146) [inline tests/test_utils.py]: Is there a reason here why you are testing the circuit twice, once with the tf backend and once with the Fock backend? The way we have the tests set up, the backend should be dynamically chosen based on an environment variable (`BACKEND=fock` or `BACKEND=tf`) in `defaults.py`, allowing us to run the same tests with different backends

Comment (ziofil) [reply tests/test_utils.py]: No, no reason, I just had not understood that your tests can do that switch automatically 😅

File: strawberryfields/utils.py (hunk 2 above)
Comment (josh146) [reply strawberryfields/utils.py]: in this function, even though the engine is returned, it is also modified directly - so even if nothing was returned, the passed engine will still be modified.

I see that further down on line 875 where it is being used, you are using deepcopy, so it should be an issue there :slightly_smiling_face:

Comment (ziofil) [reply strawberryfields/utils.py]: Yeah, I didn't want to mess up the current Engine, so I work on a deepcopy.

File: strawberryfields/utils.py (hunk 3 above)
Comment (ziofil) [reply strawberryfields/utils.py]: Done

**Comment by josh146** [comment]
Hi @ziofil, I've done a code review, as well as pushing a small commit to integrate the new functions into the SF documentation, and make sure everything renders correctly (let me know if you think anything should be changed).

It looks really fantastic! Especially the documentation, really in-depth and well covered. I think only a couple of minor commits are needed before it gets merged.

1. The comments you added to the `extract_channel` function are great, and will really help with maintainability if someone else is working on this part of the codebase in the future. Do you mind quickly adding similar comments to `extract_unitary`, or where ever you think they might be useful?

2. Tests: the test should be modified as per https://github.com/XanaduAI/strawberryfields/pull/48#discussion_r260674838 to correctly allow the environment variables to set the backends. This is something I can quickly do.

3. Documentation: It would be nice to have a short example in the documentation of usage. Another thing that crossed my mind - do you think defining equations could be useful, for example when discussing the channel representation?

**Comment by ziofil** [comment]
Hi @josh146 
1. sure, no problem, I'll do that.
2. ok, please go ahead
3. sure, how would you like the example to be formatted? Something like a doctest?

**Comment by josh146** [comment]
>     3. sure, how would you like the example to be formatted? Something like a doctest?

I was thinking something very simple, like Sphinx code-block example at the top of the Engine page in the documentation:

https://strawberryfields.readthedocs.io/en/latest/code/engine.html

>  What kind of equations do you have in mind? The definitions of channel action like
>        C(ρ) = tr_A[(ρᵀᴬ ⊗ Iᴮ)Λᴬᴮ]?

Yeah, just something like you have written, that clearly defines the various channels etc.

**Comment by ziofil** [comment]
Is the current version enough, or should I add more?

**Comment by josh146** [comment]
I was a bit late with my reply, what you have looks perfect :slightly_smiling_face:

File: tests/test_utils.py [hunk 5]
```diff
@@ -290,13 +299,381 @@ def test_random_interferometer(self):
...
+        self.assertAllAlmostEqual(final_rho_liouville, final_rho_sf, delta=self.tol)
+        self.assertAllAlmostEqual(final_rho_kraus, final_rho_sf, delta=self.tol)
+
+    def test_vectorize_unvectorize(self):
+        """Test vectorize utility function"""
+        self.logTestName()
+
+        cutoff = 4
+
+        dm = np.random.rand(*[cutoff]*8) + 1j*np.random.rand(*[cutoff]*8) # 4^8 -> (4^2)^4 -> 4^8
```
Comment (co9olguy) [inline tests/test_utils.py]: make (arbitrary) choices here instead of random

File: tests/test_utils.py [hunk 6]
```diff
@@ -290,13 +299,381 @@ def test_random_interferometer(self):
...
+    def test_extract_unitary_2_modes_fock(self):
+        """Test that unitary extraction works for 2 modes using the fock backend"""
+        self.logTestName()
+
+        eng_extract, q_extract = sf.Engine(num_subsystems=2)
+
+        S = sf.ops.Sgate(2)
+        B = sf.ops.BSgate(2.234, -1.165)
+
+        initial_state = np.complex64(np.random.rand(self.D, self.D) + 1j*np.random.rand(self.D, self.D))
```
Comment (co9olguy) [inline tests/test_utils.py]: fix choice of initial state

File: tests/test_utils.py [hunk 7]
```diff
@@ -290,13 +299,381 @@ def test_random_interferometer(self):
...
+    def test_extract_unitary_2_modes_tf(self):
+        """Test that unitary extraction works for 2 modes using the tf backend"""
+        self.logTestName()
+
+        eng_extract, q_extract = sf.Engine(num_subsystems=2)
+
+        S = sf.ops.Sgate(2)
+        B = sf.ops.BSgate(2.234, -1.165)
+
+        initial_state = np.complex64(np.random.rand(self.D, self.D) + 1j*np.random.rand(self.D, self.D))
```
Comment (co9olguy) [inline tests/test_utils.py]: fix choice of initial state

File: tests/test_utils.py [hunk 8]
```diff
@@ -290,13 +299,381 @@ def test_random_interferometer(self):
...
+    def test_extract_channel_2_modes(self):
+        """Test that channel extraction works for 2 modes"""
+        self.logTestName()
+
+        eng_extract, q_extract = sf.Engine(num_subsystems=2)
+
+        S = sf.ops.Sgate(2)
+        B = sf.ops.BSgate(2.234, -1.165)
+
+        initial_state = np.random.rand(self.D, self.D, self.D, self.D) + 1j*np.random.rand(self.D, self.D, self.D, self.D)
```
Comment (co9olguy) [inline tests/test_utils.py]: fix initial state

File: tests/test_utils.py [hunk 9]
```diff
@@ -290,13 +299,381 @@ def test_random_interferometer(self):
...
+    def test_extract_channel_2_modes_vectorize(self):
+        """Test that channel extraction works for 2 modes vectorized"""
+        self.logTestName()
+
+        eng_extract, q_extract = sf.Engine(num_subsystems=2)
+
+        S = sf.ops.Sgate(2)
+        B = sf.ops.BSgate(2.234, -1.165)
+
+        initial_state = np.random.rand(self.D, self.D, self.D, self.D) + 1j*np.random.rand(self.D, self.D, self.D, self.D)
```
Comment (co9olguy) [inline tests/test_utils.py]: fix initial state

**Comment by co9olguy** [comment]
I would commit all the minor suggestions myself, but it says I don't have push permissions to the repo 🤔

File: strawberryfields/utils.py [hunk 10]
```diff
@@ -673,3 +713,415 @@ def f_proxy(*args):
...
+        4 --|    |-- 5
+        6 --|____|-- 7
+
+    goes to
+    ::
+        (0,2) --|‾‾‾‾|-- (1,3)
+        (4,6) --|____|-- (5,7)
+
+    Args:
+        tensor (array): a tensor with :math:`4N` indices of dimension :math:`D` each
```
[the remark this answers is not in the document]
Comment (ziofil) [reply strawberryfields/utils.py]: That's not the best choice of name on my part. I just wanted to indicate that it's not acting on vectors, but it's not acting on density matrices either. Feel free to change the name (or just remove "_dm" since the name is already beginning with an underscore)

File: strawberryfields/utils.py [hunk 11]
```diff
@@ -673,3 +713,415 @@ def f_proxy(*args):
...
+
+    # here we rearrange the indices to go back to the order [in1, out1, in2, out2, etc...]
+    if backend == 'fock':
+        return np.transpose(result, [int(n) for n in np.arange(2*N).reshape((2, N)).T.reshape([-1])])
+
+    return tf.transpose(result, [int(n) for n in np.arange(2*N).reshape((2, N)).T.reshape([-1])])
+
+
+def extract_channel(engine, cutoff_dim: int, representation: str = 'choi', vectorize_modes: bool = False):
+    r"""Returns a numerical array representation of a channel from a queued Engine circuit.
```
[the remark this answers is not in the document]
Comment (ziofil) [reply strawberryfields/utils.py]: Because `extract_channel` requires a lot of index manipulations that I would have to rewrite in TensorFlow and duplicate a lot of code. in `extract_unitary` it's easier because it was just a couple of lines (which I still didn't like duplicating). 
It could be possible to use functions that have the same name in TensorFlow and Numpy, and call the one we need depending on the required backend (but it's not the best for maintainability).

File: strawberryfields/utils.py [hunk 12]
```diff
@@ -673,3 +713,415 @@ def f_proxy(*args):
...
+            otherwise it returns a rank-4N tensor, where N is the number of modes.
+
+    Returns:
+        The numerical array of the channel, according to the specified representation
+        and vectorization options.
+
+    Raises:
+        TypeError: if the gates used to construct the circuit are not all unitary or channels.
+    """
+    # if is_unitary(engine):
```
[the remark this answers is not in the document]
Comment (ziofil) [reply strawberryfields/utils.py]: I left it there as a reminder that it's a possible check that could be performed, but raising a warning seemed excessive.

File: tests/test_utils.py (hunk 8 above)
Comment (ziofil) [reply tests/test_utils.py]: What's the right way of fixing the initial array if we want it to be quite large? Here it has D^4 complex numbers (1296 as D=6)

Comment (josh146) [reply tests/test_utils.py]: You could use `np.random.seed(42)` directly above the call to `np.random.rand`. NumPy has a guarantee that it will produce the same result anywhere:

> Compatibility Guarantee A fixed seed and a fixed series of calls to ‘RandomState’ methods using the same parameters will always produce the same results up to roundoff error except when the values were incorrect.

See
1. https://stackoverflow.com/questions/40676205/cross-platform-numpy-random-seed
2. https://stackoverflow.com/questions/5836335/consistently-create-same-random-numpy-array

Comment (ziofil) [reply tests/test_utils.py]: Ah yes, that's a good idea. Is it sufficient to set it once for the whole test suite? Will it break/upset other parts of the tests?

File: strawberryfields/utils.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ziofil) [reply strawberryfields/utils.py]: Ok, I removed the "_dm" termination

File: tests/test_utils.py (hunk 5 above)
Comment (ziofil) [reply tests/test_utils.py]: I set the seed after importing numpy

File: tests/test_utils.py (hunk 6 above)
Comment (ziofil) [reply tests/test_utils.py]: I set the seed after importing numpy

File: tests/test_utils.py (hunk 7 above)
Comment (ziofil) [reply tests/test_utils.py]: I set the seed after importing numpy

File: tests/test_utils.py (hunk 8 above)
Comment (ziofil) [reply tests/test_utils.py]: I set the seed after importing numpy

File: tests/test_utils.py (hunk 9 above)
Comment (ziofil) [reply tests/test_utils.py]: I set the seed after importing numpy

File: strawberryfields/utils.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (co9olguy) [reply strawberryfields/utils.py]: No problem. It's something we can target for future enhancements, but not a deal-breaker in terms of this PR. can we add a line near the top of the docstring "Note: currently this function is only supported for the 'fock' backend"?

File: strawberryfields/utils.py [hunk 13]
```diff
@@ -673,3 +713,415 @@ def f_proxy(*args):
...
+    # here we rearrange the indices to go back to the order [in1, out1, in2, out2, etc...]
+    if backend == 'fock':
+        return np.transpose(result, [int(n) for n in np.arange(2*N).reshape((2, N)).T.reshape([-1])])
+
+    return tf.transpose(result, [int(n) for n in np.arange(2*N).reshape((2, N)).T.reshape([-1])])
+
+
+def extract_channel(engine, cutoff_dim: int, representation: str = 'choi', vectorize_modes: bool = False):
+    r"""Returns a numerical array representation of a channel from a queued Engine circuit.
+
```
Comment (co9olguy) [inline strawberryfields/utils.py]:
```suggestion
Note: currently this function is only supported for the 'fock' backend

```

File: strawberryfields/utils.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (co9olguy) [reply strawberryfields/utils.py]: Ok got it. Probably best to just remove it.

Comment (co9olguy) [inline strawberryfields/utils.py]:
```suggestion
```

File: strawberryfields/utils.py [hunk 14]
```diff
@@ -673,3 +713,415 @@ def f_proxy(*args):
...
+    Returns:
+        The numerical array of the channel, according to the specified representation
+        and vectorization options.
+
+    Raises:
+        TypeError: if the gates used to construct the circuit are not all unitary or channels.
+    """
+    # if is_unitary(engine):
+    #     #raise Warning(f"This circuit is unitary and you could use extract_unitary for a more compact representation")
```
Comment (co9olguy) [inline strawberryfields/utils.py]:
```suggestion
```

File: strawberryfields/utils.py [hunk 15]
```diff
@@ -673,3 +713,415 @@ def f_proxy(*args):
...
+    #     pass
```
Comment (co9olguy) [inline strawberryfields/utils.py]:
```suggestion
```

**Comment by co9olguy** [comment]
Just one or two tiny suggestions to commit, but all good on my side :)

File: tests/test_utils.py (hunk 8 above)
Comment (josh146) [reply tests/test_utils.py]: :+1:

Thinking about it, probably the best place to put it is in the `setUp` method of each test class.
